### PR TITLE
Generalize graphing

### LINKED ISSUE #24: Generalize Graphing
Generalize values in the app so that any trajectory formatted file can be uploaded and used

### LINKED ISSUE #47: Refactor update_figure
Right now update_figure is much too complicated. The functionality should be split up into multiple functions called by the callback. There are a few ways this could be be done. At minimum each level of for loop should be calling 1~2 functions that contain the functionality used at that level. Also there should be no logic in the definition of each go.Scatter trace. Any logic/formatting that needs to be done should be performed and stored in an appropriately named variable prior to the creation of the Scatter object.

### PR DESCRIPTION
**Pull request by alexanderkyim**
---
name: Generalize Graphing
about: Generalized graphing to handle multiple different types of data and update 

---

## Issues Addressed
Closes #24, closes #47 

## What has changed
Include each file that was changed, how it was changed (in broad terms), and why it was changed.

**tesla_ver/bubble_chart.py** was refactored to allow general graphing and handling multiple file uploads
**tesla_ver/layout.py** multiple file uploads were enabled for merging

**tesla_ver/*.py**  all other python files were just formatted with black

## Checklist of pre-requisites
All of these need to be checked in order to request a review on the Pull Request
-   [x] Does the code run?  
-   [x] Does the code follow the repository style?  
-   [x] Has Black been run on all the python files in the repository (with line length 120 `black -l 120 *`)?

## How to use the feature
`export FLASK_ENV=development` and `flask run`

## Why that change was necessary?
Previously, the graphing relied on having specific hardcoded values to run graphing.  This PR generalizes all hardcoded values to work with more

### CODE REVIEW

File: tesla_ver/bubble_chart.py [hunk 1]
```diff
@@ -14,185 +18,166 @@ def generateBubbleChart(server):
...
+            return df
+
         df = None
-        if contents is not None:
-            df = parse_contents(contents).to_json()
-        return df
+        if list_of_contents is not None:
+            df = parse_contents(list_of_contents, list_of_filenames)
+            return df.to_json()
+        return
```
Comment (DSWallach) [inline tesla_ver/bubble_chart.py:73]: Have `df = None` set at the top of the function, then assign a new value to it as necessary. There should only be one `return` statement at the end of the function, returning the appropriate value of `df`

File: tesla_ver/bubble_chart.py [hunk 2]
```diff
@@ -14,185 +18,166 @@ def generateBubbleChart(server):
...
+        Data produced by upload_data callback function"""
+        if data is None:
+            raise PreventUpdate
+        df = pd.read_json(data)
+        time_min = df["X"].min()
+        time_max = df["X"].max()
+        marks = {str(year): {"label": str(year), "style": {"visibility": "hidden"}} for year in df["X"].unique()}
+        for idx, key in enumerate(marks.keys()):
+            if idx % 4 == 0:
+                marks[key]["style"] = {"visibility": "visible"}
```
Comment (DSWallach) [inline tesla_ver/bubble_chart.py:90]: Add comments on this section

File: tesla_ver/bubble_chart.py [hunk 3]
```diff
@@ -14,185 +18,166 @@ def generateBubbleChart(server):
...
-                    legend={"x": 0, "y": 1},
-                    hovermode="closest",
-                    # Defines transition behaviors
-                    transition={"duration": 500, "easing": "cubic-in-out"},
+        actions."""
+        # Prevents updates without data
+        if not all(
+            val is not None
+            for val in [json_data, size_dropdown_name, annotation_column_name, x_column_name, y_column_name,]
+        ):
```
Comment (DSWallach) [inline tesla_ver/bubble_chart.py:149]:
```suggestion
        if None in [json_data, size_dropdown_name, annotation_column_name, x_column_name, y_column_name,]:
```

File: tesla_ver/bubble_chart.py [hunk 4]
```diff
@@ -14,185 +18,166 @@ def generateBubbleChart(server):
...
+            raise PreventUpdate
+
+        df = pd.read_json(json_data)
+        traces = list()
+        filtered_df = df[df["X"] == time_value].convert_dtypes()
+        for entity in filtered_df["alpha-3"].unique():
+            df_by_value = filtered_df[filtered_df["alpha-3"] == entity]
```
Comment (DSWallach) [inline tesla_ver/bubble_chart.py:156]: Comments on what's happening here

File: tesla_ver/bubble_chart.py (hunk 3 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart.py]: addressed in 33281c6

File: tesla_ver/bubble_chart.py (hunk 4 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart.py:171]: addressed in f9aaa7f

File: tesla_ver/bubble_chart.py [hunk 5]
```diff
@@ -1,11 +1,15 @@
...
 
-from dash.dependencies import Input, Output, State
-import plotly.graph_objs as go
 
+import dash
 import pandas as pd
+import numpy as np
+from plotly.graph_objects import Scattergl
+from dash.dependencies import Input, Output, State
+from dash.exceptions import PreventUpdate
```
[the remark this answers is not in the document]
Comment (alexanderkyim) [reply tesla_ver/bubble_chart.py:9]: addressed in f9aaa7f

File: tesla_ver/bubble_chart.py (hunk 2 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart.py]: addressed in f9aaa7f

File: tesla_ver/bubble_chart.py (hunk 1 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart.py]: addressed in fd10479

File: tesla_ver/bubble_chart.py [hunk 6]
```diff
@@ -14,185 +17,185 @@ def generateBubbleChart(server):
...
+    def upload_data(list_of_contents, list_of_filenames, _):
         """This callback handles storing the dataframe as JSON in the hidden
         component."""
+
         df = None
-        if contents is not None:
-            df = parse_contents(contents).to_json()
-        return df
+
+        if None in [list_of_contents, list_of_filenames]:
```
Comment (mstapylton) [inline tesla_ver/bubble_chart.py:30]: A better solution would check what exactly is None and display that in the error message.

File: tesla_ver/bubble_chart.py [hunk 7]
```diff
@@ -14,185 +17,185 @@ def generateBubbleChart(server):
...
+
+    @app.callback(
+        [Output("time-slider", "marks"), Output("time-slider", "min"), Output("time-slider", "max"),],
+        [Input("hidden-data", "children")],
+    )
+    def update_slider(data):
+        """This callback updates the slider values from a hidden div containing trajectory data
+        Data produced by upload_data callback function"""
+        if data is None:
+            raise PreventUpdate
```
Comment (mstapylton) [inline tesla_ver/bubble_chart.py:85]: more specific exception here

File: tesla_ver/bubble_chart.py (hunk 6 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart.py:30]: I'll address this in a later PR!

File: tesla_ver/bubble_chart.py (hunk 7 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart.py:85]: I'll address this in a later PR!

**Pull request merged**